zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js

```diff
@@ -1057,563 +966,288 @@ angular.module('zeppelinWebApp')
-        (!options.yAxis && isDiscrete(xValues)));
-
-    if (isAllDiscrete) {
-      return false;
-    }
-
-    return true;
+  function isValidSizeOption(size, data) {
+    return !!_.find(data, function(x) {return !isNaN(x[size.name])})
   };
 
-  $scope.setGraphHeight = function() {
-    var height = $('#p'+$scope.paragraph.id+'_graph').height();
+  function setGraphHeight(paragraph) {
+    var height = $('#p' + paragraph.id + '_graph').height();
 
-    var newParams = angular.copy($scope.paragraph.settings.params);
-    var newConfig = angular.copy($scope.paragraph.config);
+    var newParams = angular.copy(paragraph.settings.params);
+    var newConfig = angular.copy(paragraph.config);
 
     newConfig.graph.height = height;
 
-    commitParagraph($scope.paragraph.title, $scope.paragraph.text, newConfig, newParams);
+    commitParagraph(
+      paragraph.title, 
+      paragraph.text, 
+      newConfig, 
+      newParams
+    );
   };
 
-  /** Utility function */
-  if (typeof String.prototype.startsWith !== 'function') {
-    String.prototype.startsWith = function(str) {
-      return this.slice(0, str.length) === str;
-    };
-  }
 
-  $scope.goToSingleParagraph = function () {
+  function goToSingleParagraph(id) {
     var noteId = $route.current.pathParams.noteId;
-    var redirectToUrl = location.protocol + '//' + location.host + '/#/notebook/' + noteId + '/paragraph/' + $scope.paragraph.id+'?asIframe';
+    var redirectToUrl = location.protocol + 
+      '//' + 
+      location.host + 
+      '/#/notebook/' + 
+      noteId + 
+      '/paragraph/' + 
+      id + 
+      '?asIframe';
     $window.open(redirectToUrl);
   };
+
+  vm.aceLoaded = aceLoaded
+  vm.goToSingleParagraph = goToSingleParagraph
+  vm.setGraphHeight = setGraphHeight
+  vm.isValidSizeOption = isValidSizeOption
+  vm.removeGraphOptionKeys = removeGraphOptionKeys
+  vm.removeGraphOptionValues = removeGraphOptionValues
+  vm.removeGraphOptionGroups = removeGraphOptionGroups
+  vm.setGraphOptionValueAggr = setGraphOptionValueAggr
+  vm.removeScatterOptionXaxis = removeScatterOptionXaxis
+  vm.removeScatterOptionYaxis = removeScatterOptionYaxis
+  vm.removeScatterOptionGroup = removeScatterOptionGroup
+  vm.removeScatterOptionSize = removeScatterOptionSize
+  vm.onGraphOptionChange = onGraphOptionChange
+  vm.isGraphMode = isGraphMode
+  vm.setGraphMode = setGraphMode
+  vm.getGraphMode = getGraphMode
+  vm.getBase64ImageSrc = getBase64ImageSrc
+  vm.getResultType = getResultType
+  vm.getExecutionTime = getExecutionTime
+  vm.getProgress = getProgress
+  vm.getEditorValue = getEditorValue
+  vm.loadForm = loadForm
+  vm.toggleOutput = toggleOutput
+  vm.toggleGraphOption = toggleGraphOption
+  vm.changeColWidth = changeColWidth
+  vm.setColumnWidthClass = setColumnWidthClass
+  vm.showTitle = showTitle
+  vm.hideTitle = hideTitle
+  vm.setTitle = setTitle
+  vm.toggleEditor = toggleEditor
+  vm.removeParagraph = removeParagraph
+  vm.moveUp = moveUp
```

Comment:
Didn't know you could access functions outside of the controller scope from the view